install/0000_90_cluster-version-operator_02_servicemonitor.yaml

```diff
@@ -16,3 +16,49 @@ spec:
   selector:
     matchLabels:
       k8s-app: cluster-version-operator
+---
+apiVersion: monitoring.coreos.com/v1
+kind: PrometheusRule
+metadata:
+  labels:
+    k8s-app: cluster-version-operator
+  name: cluster-version-operator
+  namespace: openshift-cluster-version
+spec:
+  groups:
+  - name: cluster-version
+    rules:
+    - alert: ClusterVersionOperatorDown
+      annotations:
+        message: Cluster version operator has disappeared from Prometheus target discovery. Operator may be down or disabled, cluster will not be kept up to date and upgrades will not be possible.
+      expr: |
+        absent(up{job="cluster-version-operator"} == 1)
+      for: 15m
+      labels:
+        severity: critical
+  - name: cluster-operators
+    rules:
+    - alert: ClusterOperatorDown
+      annotations:
+        message: Cluster operator {{ "{{ $labels.name }}" }} has not been available for 10 mins. Operator may be down or disabled, cluster will not be kept up to date and upgrades will not be possible.
```

Comment:
Why this alert as opposed to an alert like `ClusterVersionOperatorDown` for each cluster operator? As I understand this alert has multiple indirections which we want to avoid at all costs in monitoring.

Reply:
Cluster version operator is not a SLO. it does not have a cluster operator. that's why the `ClusterVersionOperatorDown` right above.

Reply:
> As I understand this alert has multiple indirections which we want to avoid at all costs in monitoring.

hmm, do you mean this alert can point to as many cluster operators.

Reply:
No, it's an alert on a metric that is not exposed by the process of origin. As in the cluster-version-operator is exposing a metric on behalf of another operator, and even worse, with information provided by the other operator. This requires that _everything_ is working properly to provide a failure case. This is why metrics should be exposed whereever the origin of any information is.

Reply:
We are monitoring the core control loop of the cluster.  We have to monitor both, not just one or the other.  This alert reflects the “making progress” part of the product - operator A and B coordinate via this api, not via alerts and metrics.  We *should* have per operator monitoring.  We *must* have core product control loop monitoring and alerting.

Reply:
If we move forward with this PR, then I would want to see *concrete* actions as to how we will properly resolve this in the very next release (4.3), which would be the _requirement_ for each cluster-operator to:

* expose identify and expose metrics to describe their SLIs
* only be allowed to page customers (that includes our own SRE) with alerts that describe error budget is burnt at a high rate
* not include alerts that are meant for our engineering team's, these are to be done via telemetry infrastructure (as in alert on data in telemetry)

Reply:
Degraded operators are functional indistinguishable from broken clusters.  We have alert because we know something is wrong.

Evidence wise I haven’t seen more than a small percentage of degraded operators that were not error conditions that should result in a customer requesting a bug fix for users.

For disconnected users alerts are mandatory on cluster to inform them to contact red hat.